pkg/gardenlet/controller/vpaevictionrequirements/reconciler.go

```diff
@@ -0,0 +1,164 @@
+	log := logf.FromContext(ctx)
+	seedCtx, cancel := controllerutils.GetMainReconciliationContext(ctx, controllerutils.DefaultReconciliationTimeout)
+	defer cancel()
+
+	vpa := &vpaautoscalingv1.VerticalPodAutoscaler{}
+	err := r.SeedClient.Get(seedCtx, request.NamespacedName, vpa)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	log.Info("Reconciling vpa", "namespace/name", request.NamespacedName)
+
+	// double check just for fun: does the vpa have our label?
+	if metav1.HasLabel(vpa.ObjectMeta, constants.LabelVPAEvictionRequirementDownscaleInMaintenanceOnly) {
+		log.Info("Found the label " + constants.LabelVPAEvictionRequirementDownscaleInMaintenanceOnly)
+		return r.reconcileVPAForDownscaleInMaintenanceOnly(seedCtx, vpa)
+	} else if metav1.HasLabel(vpa.ObjectMeta, constants.LabelVPAEvictionRequirementDownscaleDisabled) {
+		log.Info("Found the label " + constants.LabelVPAEvictionRequirementDownscaleDisabled)
+		return r.reconcileVPAForDownscaleDisabled(seedCtx, vpa)
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *Reconciler) reconcileVPAForDownscaleInMaintenanceOnly(seedCtx context.Context, vpa *vpaautoscalingv1.VerticalPodAutoscaler) (reconcile.Result, error) {
+	log := logf.FromContext(seedCtx)
+	if !metav1.HasAnnotation(vpa.ObjectMeta, constants.AnnotationShootMaintenanceWindow) {
+		err := errors.New("didn't find maintenance window annotation on VPA, but VPA had label to be downscaled in maintenance only")
+		return reconcile.Result{}, err
+	}
+
+	windowAnnotation := vpa.GetAnnotations()[constants.AnnotationShootMaintenanceWindow]
+	splitWindowAnnotation := strings.Split(windowAnnotation, ",")
+	maintenanceTimeWindow, err := timewindow.ParseMaintenanceTimeWindow(splitWindowAnnotation[0], splitWindowAnnotation[1])
+	if err != nil {
+		log.Error(err, "Error during parsing the maintenance window from annotation", "begin", splitWindowAnnotation[0], "end", splitWindowAnnotation[1])
+		return reconcile.Result{}, err
+	}
+	isNowInMaintenanceTimeWindow := maintenanceTimeWindow.Contains(r.Clock.Now())
+
+	if isNowInMaintenanceTimeWindow {
+		log.Info("Shoot is inside maintenance window, removing the EvictionRequirement to allow downscaling", "shoot-namespace", vpa.GetNamespace(), "maintenanceWindow", maintenanceTimeWindow)
+		_, err := controllerutils.GetAndCreateOrMergePatch(seedCtx, r.SeedClient, vpa, func() error {
+			if !hasUpscaleOnlyEvictionRequirement(vpa.Spec.UpdatePolicy.EvictionRequirements) {
+				return nil
+			}
+			vpa.Spec.UpdatePolicy.EvictionRequirements = []*vpaautoscalingv1.EvictionRequirement{}
+			return nil
+		})
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// requeue when the maintenance window ends, such that we can add the EvictionRequirement again
+		endTime := maintenanceTimeWindow.AdjustedEnd(r.Clock.Now())
+		requeueAfter := endTime.Sub(r.Clock.Now())
+		log.Info("Requeueing VPA", "requeueAfter", requeueAfter)
```

Comment:
The log message has no identifying component. Please either pass in namespace/name or create the logger beforehand to always include those.

Reply:
done.